### PR TITLE
exoscale: terraform setup for k8s WIP

### PR DESCRIPTION
**Pull request by greut**

### CODE REVIEW

**Comment by greut** [comment]
`hosts` is a python script parsing the `terraform.tfstate` into the inventory.

```console
$ ansible -i hosts -m ping all
example-k8s-master-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
example-k8s-master-ne-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
example-etcd-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
example-k8s-node-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
```

File: contrib/terraform/exoscale/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+Provision a Kubernetes cluster with Terraform on Exoscale.
+
+## Status
+
+This will install a Kubernetes cluster on the Exocale Cloud provider.
+
+## Approach
+
+The Terraform configuration inspeccts variables found in [variables.tf](variables.tf) to create the resources. There is a [python script](../terraform.py) that reads the generated `.tfstate` file to build the dynamic inventory that is consumed by the main ansible script to actually install Kubernetes.
```
Comment (marcaurele) [inline contrib/terraform/exoscale/README.md]: typo: inspects

**Comment by greut** [comment]
```
...

TASK [kubespray-defaults : Configure defaults] *******************************************************************************
Thursday 22 March 2018  12:33:49 +0100 (0:00:01.294)       0:01:01.008 ********
fatal: [exokube-k8s-node-1]: FAILED! => {"msg": "{'http_proxy': \"{{ http_proxy| default ('') }}\", 'https_proxy': \"{{ https_proxy| default ('') }}\", 'no_proxy': '{{ no_proxy }}'}: {%- if loadbalancer_apiserver is defined -%} {{ apiserver_loadbalancer_domain_name| default('') }}, {{ loadbalancer_apiserver.address | default('') }}, {%- endif -%} {%- for item in (groups['k8s-cluster'] + groups['etcd'] + groups['calico-rr']|default([]))|unique -%} {{ hostvars[item]['access_ip'] | default(hostvars[item]['ip'] | default(hostvars[item]['ansible_default_ipv4']['address'])) }}, {%-   if (item != hostvars[item]['ansible_hostname']) -%} {{ hostvars[item]['ansible_hostname'] }}, {{ hostvars[item]['ansible_hostname'] }}.{{ dns_domain }}, {%-   endif -%} {{ item }},{{ item }}.{{ dns_domain }}, {%- endfor -%} 127.0.0.1,localhost: The ipaddr filter requires python-netaddr be installed on the ansible controller"}
fatal: [exokube-k8s-master-ne-2]: FAILED! => {"msg": "{'http_proxy': \"{{ http_proxy| default ('') }}\", 'https_proxy': \"{{ https_proxy| default ('') }}\", 'no_proxy': '{{ no_proxy }}'}: {%- if loadbalancer_apiserver is defined -%} {{ apiserver_loadbalancer_domain_name| default('') }}, {{ loadbalancer_apiserver.address | default('') }}, {%- endif -%} {%- for item in (groups['k8s-cluster'] + groups['etcd'] + groups['calico-rr']|default([]))|unique -%} {{ hostvars[item]['access_ip'] | default(hostvars[item]['ip'] | default(hostvars[item]['ansible_default_ipv4']['address'])) }}, {%-   if (item != hostvars[item]['ansible_hostname']) -%} {{ hostvars[item]['ansible_hostname'] }}, {{ hostvars[item]['ansible_hostname'] }}.{{ dns_domain }}, {%-   endif -%} {{ item }},{{ item }}.{{ dns_domain }}, {%- endfor -%} 127.0.0.1,localhost: The ipaddr filter requires python-netaddr be installed on the ansible controller"}
fatal: [exokube-k8s-master-2]: FAILED! => {"msg": "{'http_proxy': \"{{ http_proxy| default ('') }}\", 'https_proxy': \"{{ https_proxy| default ('') }}\", 'no_proxy': '{{ no_proxy }}'}: {%- if loadbalancer_apiserver is defined -%} {{ apiserver_loadbalancer_domain_name| default('') }}, {{ loadbalancer_apiserver.address | default('') }}, {%- endif -%} {%- for item in (groups['k8s-cluster'] + groups['etcd'] + groups['calico-rr']|default([]))|unique -%} {{ hostvars[item]['access_ip'] | default(hostvars[item]['ip'] | default(hostvars[item]['ansible_default_ipv4']['address'])) }}, {%-   if (item != hostvars[item]['ansible_hostname']) -%} {{ hostvars[item]['ansible_hostname'] }}, {{ hostvars[item]['ansible_hostname'] }}.{{ dns_domain }}, {%-   endif -%} {{ item }},{{ item }}.{{ dns_domain }}, {%- endfor -%} 127.0.0.1,localhost: The ipaddr filter requires python-netaddr be installed on the ansible controller"}
fatal: [exokube-etcd-2]: FAILED! => {"msg": "{'http_proxy': \"{{ http_proxy| default ('') }}\", 'https_proxy': \"{{ https_proxy| default ('') }}\", 'no_proxy': '{{ no_proxy }}'}: {%- if loadbalancer_apiserver is defined -%} {{ apiserver_loadbalancer_domain_name| default('') }}, {{ loadbalancer_apiserver.address | default('') }}, {%- endif -%} {%- for item in (groups['k8s-cluster'] + groups['etcd'] + groups['calico-rr']|default([]))|unique -%} {{ hostvars[item]['access_ip'] | default(hostvars[item]['ip'] | default(hostvars[item]['ansible_default_ipv4']['address'])) }}, {%-   if (item != hostvars[item]['ansible_hostname']) -%} {{ hostvars[item]['ansible_hostname'] }}, {{ hostvars[item]['ansible_hostname'] }}.{{ dns_domain }}, {%-   endif -%} {{ item }},{{ item }}.{{ dns_domain }}, {%- endfor -%} 127.0.0.1,localhost: The ipaddr filter requires python-netaddr be installed on the ansible controller"}
fatal: [exokube-etcd-1]: FAILED! => {"msg": "{'http_proxy': \"{{ http_proxy| default ('') }}\", 'https_proxy': \"{{ https_proxy| default ('') }}\", 'no_proxy': '{{ no_proxy }}'}: {%- if loadbalancer_apiserver is defined -%} {{ apiserver_loadbalancer_domain_name| default('') }}, {{ loadbalancer_apiserver.address | default('') }}, {%- endif -%} {%- for item in (groups['k8s-cluster'] + groups['etcd'] + groups['calico-rr']|default([]))|unique -%} {{ hostvars[item]['access_ip'] | default(hostvars[item]['ip'] | default(hostvars[item]['ansible_default_ipv4']['address'])) }}, {%-   if (item != hostvars[item]['ansible_hostname']) -%} {{ hostvars[item]['ansible_hostname'] }}, {{ hostvars[item]['ansible_hostname'] }}.{{ dns_domain }}, {%-   endif -%} {{ item }},{{ item }}.{{ dns_domain }}, {%- endfor -%} 127.0.0.1,localhost: The ipaddr filter requires python-netaddr be installed on the ansible controller"}

NO MORE HOSTS LEFT ***********************************************************************************************************
        to retry, use: --limit @/home/yoan/soft/kubespray/cluster.retry

PLAY RECAP *******************************************************************************************************************
exokube-etcd-1             : ok=11   changed=1    unreachable=0    failed=1   
exokube-etcd-2             : ok=11   changed=1    unreachable=0    failed=1   
exokube-k8s-master-2       : ok=11   changed=1    unreachable=0    failed=1   
exokube-k8s-master-ne-2    : ok=11   changed=1    unreachable=0    failed=1   
exokube-k8s-node-1         : ok=11   changed=1    unreachable=0    failed=1   
localhost                  : ok=2    changed=0    unreachable=0    failed=0

Thursday 22 March 2018  12:33:49 +0100 (0:00:00.229)       0:01:01.237 ********
===============================================================================
bootstrap-os : Bootstrap | Install python 2.x and pip ---------------------------------------------------------------- 37.94s
gather facts from all instances --------------------------------------------------------------------------------------- 6.40s
bootstrap-os : Assign inventory name to unconfigured hostnames (non-CoreOS) ------------------------------------------- 2.20s
bootstrap-os : Gather nodes hostnames --------------------------------------------------------------------------------- 2.18s
Gathering Facts ------------------------------------------------------------------------------------------------------- 1.55s
bootstrap-os : Bootstrap | Check if bootstrap is needed --------------------------------------------------------------- 1.44s
download : Sync container --------------------------------------------------------------------------------------------- 1.29s
bootstrap-os : check if atomic host ----------------------------------------------------------------------------------- 1.25s
bootstrap-os : Remove require tty ------------------------------------------------------------------------------------- 1.24s
download : Download items --------------------------------------------------------------------------------------------- 1.20s
download : Download items --------------------------------------------------------------------------------------------- 0.38s
download : Sync container --------------------------------------------------------------------------------------------- 0.33s
bootstrap-os : set_fact ----------------------------------------------------------------------------------------------- 0.25s
kubespray-defaults : Configure defaults ------------------------------------------------------------------------------- 0.23s
bootstrap-os : Bootstrap | Check if bootstrap is needed --------------------------------------------------------------- 0.20s
download : Download items --------------------------------------------------------------------------------------------- 0.20s
bootstrap-os : set_fact ----------------------------------------------------------------------------------------------- 0.19s
kubespray-defaults : Configure defaults ------------------------------------------------------------------------------- 0.16s
download : Sync container --------------------------------------------------------------------------------------------- 0.15s
download : include_tasks ---------------------------------------------------------------------------------------------- 0.15s
```

**Comment by greut** [comment]
`pip install ansible netaddr` and it gets a little bit better.

```
fatal: [exokube-k8s-master-2]: FAILED! => {"changed": false, "cmd": "apt-get install python-apt -y -q", "msg": "E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4.0~beta3build2_amd64.deb  Temporary failure resolv
ing 'de.archive.ubuntu.com'\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?", "rc": 100, "stderr": "E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4.0~beta3build2_amd64.de
b  Temporary failure resolving 'de.archive.ubuntu.com'\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?\n", "stderr_lines": ["E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1
.4.0~beta3build2_amd64.deb  Temporary failure resolving 'de.archive.ubuntu.com'", "E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state in
formation...\nThe following package was automatically installed and is no longer required:\n  grub-pc-bin\nUse 'sudo apt autoremove' to remove it.\nSuggested packages:\n  python-apt-dbg python-apt-doc\nThe following NEW packages will be installed:\n  pyt
hon-apt\n0 upgraded, 1 newly installed, 0 to remove and 22 not upgraded.\nNeed to get 143 kB of archives.\nAfter this operation, 647 kB of additional disk space will be used.\nErr:1 http://de.archive.ubuntu.com/ubuntu artful/main amd64 python-apt amd64 1
.4.0~beta3build2\n  Temporary failure resolving 'de.archive.ubuntu.com'\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following package was automatically installed and is no longer re
quired:", "  grub-pc-bin", "Use 'sudo apt autoremove' to remove it.", "Suggested packages:", "  python-apt-dbg python-apt-doc", "The following NEW packages will be installed:", "  python-apt", "0 upgraded, 1 newly installed, 0 to remove and 22 not upgrad
ed.", "Need to get 143 kB of archives.", "After this operation, 647 kB of additional disk space will be used.", "Err:1 http://de.archive.ubuntu.com/ubuntu artful/main amd64 python-apt amd64 1.4.0~beta3build2", "  Temporary failure resolving 'de.archive.u
buntu.com'"]}
fatal: [exokube-k8s-master-ne-2]: FAILED! => {"changed": false, "cmd": "apt-get install python-apt -y -q", "msg": "E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4.0~beta3build2_amd64.deb  Temporary failure res
olving 'de.archive.ubuntu.com'\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?", "rc": 100, "stderr": "E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4.0~beta3build2_amd64
.deb  Temporary failure resolving 'de.archive.ubuntu.com'\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?\n", "stderr_lines": ["E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-ap
t_1.4.0~beta3build2_amd64.deb  Temporary failure resolving 'de.archive.ubuntu.com'", "E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state
 information...\nThe following package was automatically installed and is no longer required:\n  grub-pc-bin\nUse 'sudo apt autoremove' to remove it.\nSuggested packages:\n  python-apt-dbg python-apt-doc\nThe following NEW packages will be installed:\n  
python-apt\n0 upgraded, 1 newly installed, 0 to remove and 22 not upgraded.\nNeed to get 143 kB of archives.\nAfter this operation, 647 kB of additional disk space will be used.\nErr:1 http://de.archive.ubuntu.com/ubuntu artful/main amd64 python-apt amd6
4 1.4.0~beta3build2\n  Temporary failure resolving 'de.archive.ubuntu.com'\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following package was automatically installed and is no longer
 required:", "  grub-pc-bin", "Use 'sudo apt autoremove' to remove it.", "Suggested packages:", "  python-apt-dbg python-apt-doc", "The following NEW packages will be installed:", "  python-apt", "0 upgraded, 1 newly installed, 0 to remove and 22 not upg
raded.", "Need to get 143 kB of archives.", "After this operation, 647 kB of additional disk space will be used.", "Err:1 http://de.archive.ubuntu.com/ubuntu artful/main amd64 python-apt amd64 1.4.0~beta3build2", "  Temporary failure resolving 'de.archiv
e.ubuntu.com'"]}
fatal: [exokube-k8s-node-1]: FAILED! => {"changed": false, "cmd": "apt-get install python-apt -y -q", "msg": "E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4.0~beta3build2_amd64.deb  Temporary failure resolvin
g 'de.archive.ubuntu.com'\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?", "rc": 100, "stderr": "E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4.0~beta3build2_amd64.deb 
 Temporary failure resolving 'de.archive.ubuntu.com'\nE: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?\n", "stderr_lines": ["E: Failed to fetch http://de.archive.ubuntu.com/ubuntu/pool/main/p/python-apt/python-apt_1.4
.0~beta3build2_amd64.deb  Temporary failure resolving 'de.archive.ubuntu.com'", "E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state info
rmation...\nThe following package was automatically installed and is no longer required:\n  grub-pc-bin\nUse 'sudo apt autoremove' to remove it.\nSuggested packages:\n  python-apt-dbg python-apt-doc\nThe following NEW packages will be installed:\n  pytho
n-apt\n0 upgraded, 1 newly installed, 0 to remove and 22 not upgraded.\nNeed to get 143 kB of archives.\nAfter this operation, 647 kB of additional disk space will be used.\nErr:1 http://de.archive.ubuntu.com/ubuntu artful/main amd64 python-apt amd64 1.4
.0~beta3build2\n  Temporary failure resolving 'de.archive.ubuntu.com'\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following package was automatically installed and is no longer requ
ired:", "  grub-pc-bin", "Use 'sudo apt autoremove' to remove it.", "Suggested packages:", "  python-apt-dbg python-apt-doc", "The following NEW packages will be installed:", "  python-apt", "0 upgraded, 1 newly installed, 0 to remove and 22 not upgraded
.", "Need to get 143 kB of archives.", "After this operation, 647 kB of additional disk space will be used.", "Err:1 http://de.archive.ubuntu.com/ubuntu artful/main amd64 python-apt amd64 1.4.0~beta3build2", "  Temporary failure resolving 'de.archive.ubu
ntu.com'"]}
```

**Comment by greut** [comment]
```
$ more /etc/resolv.conf
# Ansible entries BEGIN
domain cluster.local
search default.svc.cluster.local svc.cluster.local
nameserver 
options ndots:2
options timeout:2
options attempts:2
```

**Comment by greut** [comment]
```
$ ansible -i hosts -m ping all
exo-etcd-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
exo-etcd-2 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
exo-k8s-node-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}
exo-k8s-master-1 | SUCCESS => {
    "changed": false,
    "ping": "pong"
}

$ cd ../..
$ ansible-playbook --become -i inventory/$CLUSTER/hosts cluster.yml
```

... 10 minutes later ...

```
Tuesday 27 March 2018  17:50:12 +0200 (0:00:00.038)       0:09:16.711 ********* 
=============================================================================== 
bootstrap-os : Bootstrap | Install python 2.x and pip ---------------------------------------------------------------- 32.78s
kubernetes/master : kubeadm | Initialize first master ---------------------------------------------------------------- 29.75s
etcd : Configure | Join member(s) to etcd-events cluster one at a time ----------------------------------------------- 20.17s
etcd : Configure | Join member(s) to etcd cluster one at a time ------------------------------------------------------ 20.17s
etcd : Configure | Join member(s) to etcd cluster one at a time ------------------------------------------------------ 20.16s
etcd : Configure | Join member(s) to etcd-events cluster one at a time ----------------------------------------------- 20.15s
download : container_download | Download containers if pull is required or told to always pull (all nodes) ----------- 18.84s
docker : Docker | pause while Docker restarts ------------------------------------------------------------------------ 10.19s
docker : ensure docker packages are installed ------------------------------------------------------------------------- 9.64s
download : container_download | Download containers if pull is required or told to always pull (all nodes) ------------ 9.31s
download : container_download | Download containers if pull is required or told to always pull (all nodes) ------------ 7.51s
kubernetes-apps/ansible : Kubernetes Apps | Lay Down KubeDNS Template ------------------------------------------------- 6.63s
kubernetes/preinstall : Install packages requirements ----------------------------------------------------------------- 5.49s
kubernetes-apps/ansible : Kubernetes Apps | Start Resources ----------------------------------------------------------- 5.38s
download : container_download | Download containers if pull is required or told to always pull (all nodes) ------------ 5.35s
kubernetes-apps/ansible : Kubernetes Apps | Delete old KubeDNS resources ---------------------------------------------- 5.21s
kubernetes/preinstall : Update package management cache (APT) --------------------------------------------------------- 5.01s
gather facts from all instances --------------------------------------------------------------------------------------- 5.00s
docker : ensure docker-engine repository is enabled ------------------------------------------------------------------- 4.95s
network_plugin/calico : Calico | Create calico manifests -------------------------------------------------------------- 4.94s
```

**Comment by greut** [comment]
```
$ kubectl get nodes
NAME               STATUS    ROLES     AGE       VERSION
exo-k8s-master-1   Ready     master    4m        v1.9.3+coreos.0
exo-k8s-node-1     Ready     <none>    3m        v1.9.3+coreos.0
```

**Pull request closed without merging**